Lib9c/Action/CombinationEquipment.cs

```diff
@@ -255,65 +259,117 @@ public override IAccountStateDelta Execute(IActionContext context)
+                    dailyCostState.Count++;
+                    dailyCostState.CRYSTAL += costCrystal;
+                    // Update Weekly Formula.
+                    weeklyCostState.Count++;
+                    weeklyCostState.CRYSTAL += costCrystal;
+
+                    var crystalBalance =
+                        states.GetBalance(context.Signer, CrystalCalculator.CRYSTAL);
+                    if (costCrystal > crystalBalance)
+                    {
+                        throw new NotEnoughFungibleAssetValueException(
+                            $"required {costCrystal}, but balance is {crystalBalance}");
+                    }
+
+                    states = states
+                        .SetState(dailyCostState.Address, dailyCostState.Serialize())
+                        .SetState(weeklyCostState.Address, weeklyCostState.Serialize())
+                        .TransferAsset(context.Signer, Addresses.MaterialCost, costCrystal);
                 }
-            }
-            // ~Remove Required Materials
-            if (costCrystal > 0 * CrystalCalculator.CRYSTAL)
-            {
-                var crystalFluctuationSheet = sheets.GetSheet<CrystalFluctuationSheet>();
-                var row = crystalFluctuationSheet.Values
-                    .First(r => r.Type == CrystalFluctuationSheet.ServiceType.Combination);
-                var (dailyCostState, weeklyCostState, _, _) = states.GetCrystalCostStates(context.BlockIndex, row.BlockInterval);
-                // 1x fixed crystal cost.
-                costCrystal = CrystalCalculator.CalculateCombinationCost(costCrystal, row: row, prevWeeklyCostState: null, beforePrevWeeklyCostState: null);
-                // Update Daily Formula.
-                dailyCostState.Count++;
-                dailyCostState.CRYSTAL += costCrystal;
-                // Update Weekly Formula.
-                weeklyCostState.Count++;
-                weeklyCostState.CRYSTAL += costCrystal;
 
-                var crystalBalance = states.GetBalance(context.Signer, CrystalCalculator.CRYSTAL);
-                if (costCrystal > crystalBalance)
+                int gainHammerPoint;
+                if (!subRecipeId.HasValue || recipeRow.SubRecipeIds[0] == subRecipeId.Value)
                 {
-                    throw new NotEnoughFungibleAssetValueException($"required {costCrystal}, but balance is {crystalBalance}");
+                    gainHammerPoint = 1;
                 }
-
-                states = states
-                    .SetState(dailyCostState.Address, dailyCostState.Serialize())
-                    .SetState(weeklyCostState.Address, weeklyCostState.Serialize())
-                    .TransferAsset(context.Signer, Addresses.MaterialCost, costCrystal);
+                else
+                {
+                    gainHammerPoint = 2;
+                }
+                hammerPointState.AddHammerPoint(gainHammerPoint, sheets.GetSheet<CrystalHammerPointSheet>());
```

Comment:
I think you can use [l.267](https://github.com/planetarium/lib9c/blob/863100e735685056226ae7de56f4f2d1ab58dc31/Lib9c/Action/CombinationEquipment.cs#L267) instead of `sheets.GetSheet<CrystalHammerPointSheet>()`